File: web/src/pesosPropios.ts
import type { Ejercicio } from "./datos";

/**
 * Los pesos y notas que el entrenador le puso a ESTA clienta, encima de los de la plantilla que
 * comparte con las demás.
 *
 * GEMELO: `domain/PesosPropios.kt` en Kotlin. Si cambia allá, cambia acá.
 *
 * Se indexan por nombre de ejercicio y no por posición para que reordenar la plantilla no le
 * despegue los pesos a nadie. Sólo se aplican mientras siga una plantilla: en rutina propia la
 * verdad es su copia, y aplicarlos encima pisaría lo que el entrenador escribió en su editor.
 */
export function claveEjercicio(nombre: string): string {
  return nombre.trim().toLowerCase().replace(/\s+/g, " ");
}

export function conPesosPropios(
  ejercicios: Ejercicio[],
  pesosPropios: Record<string, string> | undefined
): Ejercicio[] {
  // Firestore omite los campos que nunca se escribieron, así que una clienta a la que nunca se
  // le puso un peso propio llega sin el mapa, no con uno vacío.
  if (!pesosPropios) return ejercicios;
  return ejercicios.map((e) => {
    const propio = pesosPropios[claveEjercicio(e.nombre ?? "")];
    // Un valor en blanco no borra el de la plantilla, cae a él: vaciar el campo es la forma de
    // decir "usa el del grupo".
    return propio && propio.trim() !== "" ? { ...e, pesoONota: propio } : e;
  });
}

File: app/src/main/java/com/osfit/app/domain/PesosPropios.kt
package com.osfit.app.domain

import com.osfit.app.data.model.Ejercicio

/**
 * Los pesos y notas que el entrenador le puso a **una clienta concreta**, encima de los de la
 * plantilla que comparte con las demás.
 *
 * El problema que resuelven: `pesoONota` vive dentro del `Ejercicio` de la plantilla, así que
 * es el mismo para todas las que la siguen. El ejercicio es del programa, pero **el peso es de
 * la persona**, y hasta el 2026-09-15 tener pesos propios obligaba a sacarla de la plantilla —
 * y entonces dejaban de llegarle los cambios. Con esto se queda en la plantilla y sus pesos
 * viven en su documento.
 *
 * **Se indexan por nombre de ejercicio y no por posición** para que reordenar los ejercicios de
 * la plantilla, o insertar uno en medio, no le despegue los pesos a nadie. El costo es que
 * renombrar un ejercicio pierde su peso propio: no se rompe nada, simplemente vuelve a mostrar
 * el de la plantilla.
 *
 * **Sólo se aplican mientras siga una plantilla** (`plantillaOrigenId` con valor). En rutina
 * propia la verdad es su copia, y aplicarlos encima pisaría lo que el entrenador acaba de
 * escribir en su editor.
 */

/**
 * La forma en que se guarda el nombre para buscarlo: sin espacios de sobra, sin mayúsculas.
 * Así "Press Banca", "press banca" y "  Press  banca " son el mismo ejercicio, que es como los
 * escribe una persona con prisa.
 *
 * GEMELO: `claveEjercicio` en `web/src/pesosPropios.ts`. Si cambia acá, cambia allá.
 */
fun claveEjercicio(nombre: String): String =
    nombre.trim().lowercase().replace(Regex("\\s+"), " ")

/**
 * Cambia el `pesoONota` de cada ejercicio por el propio de la clienta, si tiene uno.
 *
 * Un valor en blanco no borra el de la plantilla, cae a él: dejar el campo vacío es la forma de
 * decir "usa el del grupo", y guardar el vacío sólo llenaría el mapa de entradas muertas.
 *
 * GEMELO: `conPesosPropios` en `web/src/pesosPropios.ts`.
 */
fun conPesosPropios(
    ejercicios: List<Ejercicio>,
    pesosPropios: Map<String, String>
): List<Ejercicio> {
    if (pesosPropios.isEmpty()) return ejercicios
    return ejercicios.map { ejercicio ->
        val propio = pesosPropios[claveEjercicio(ejercicio.nombre)]
        if (propio.isNullOrBlank()) ejercicio else ejercicio.copy(pesoONota = propio)
    }
}

/**
 * El mapa con [nombre] puesto en [valor], listo para guardar. Un valor en blanco **quita** la
 * entrada en vez de guardarla vacía, para que el mapa no acumule claves de ejercicios que ya
 * nadie usa y para que borrar el campo signifique "vuelve al de la plantilla".
 */
fun conPesoPropio(
    pesosPropios: Map<String, String>,
    nombre: String,
    valor: String
): Map<String, String> {
    val clave = claveEjercicio(nombre)
    if (clave.isBlank()) return pesosPropios
    return if (valor.isBlank()) {
        pesosPropios - clave
    } else {
        pesosPropios + (clave to valor)
    }
}
